simulator/property_collector.go

```diff
@@ -214,6 +218,64 @@ func fieldValue(rval reflect.Value, p string) (interface{}, error) {
 	return value, nil
 }
 
+func fieldValueKey(rval reflect.Value, p mo.Field) (interface{}, error) {
+	if rval.Kind() != reflect.Slice {
+		return nil, errInvalidField
+	}
+
+	zero := reflect.Value{}
+
+	for i := 0; i < rval.Len(); i++ {
+		item := rval.Index(i)
+		if item.Kind() == reflect.Interface {
+			item = item.Elem()
+		}
+		if item.Kind() == reflect.Ptr {
+			item = item.Elem()
+		}
+		if item.Kind() != reflect.Struct {
+			return reflect.Value{}, errInvalidField
+		}
+
+		field := item.FieldByName("Key")
```

Comment:
I am having trouble tracing how we get here, so maybe this is a non-issue, but what if the property path is for an array of non-struct types, such as an array of integers or strings? I believe a numerical index should work in that case.

Reply:
afaik, property path array index can only be used when struct has a `key` field, see for example:

```console
% govc object.collect $GOVC_VM disabledMethod
disabledMethod  []string  MakePrimaryVM_Task,TerminateFaultTolerantVM_Task,...

% govc object.collect $GOVC_VM 'disabledMethod[1]'
govc: ServerFaultCode: InvalidProperty
```

Another case, AuthorizationRole has no `key` field. And while it does have a unique `roleId` field, same `InvalidProperty` result when using a valid `roleId`
```console
% id=$(govc object.collect -o -json AuthorizationManager:AuthorizationManager roleList | jq .roleList[].roleId | head -1)

% govc object.collect AuthorizationManager:AuthorizationManager "roleList[$id]"
govc: ServerFaultCode: InvalidProperty
```

This [doc](https://docs.vmware.com/en/VMware-vSphere/8.0/vsphere-web-services-programming-guide/GUID-DB836CA6-8316-4B2F-8CE5-8AD933F8DE36.html) also seems to indicate must be key-based:

> Nested properties can also refer to entries in a key-based array